File: PASCAL/automato.pas
unit automato;

{$mode objfpc}{$H+}

interface

uses
  SysUtils, Classes, Generics.Collections, tipos;

type
  TTabela = specialize TDictionary<String, TStringSet>;

function ConstruirTabelaTransicoes(const A: TAutomato): TTabela;
function FechoLambda(Estados: TStringSet; Tabela: TTabela): TStringSet;
function AceitaPalavra(const A: TAutomato; const Palavra: String): Boolean;
procedure LiberarTabela(var T: TTabela);

implementation

function ChaveTransicao(const Estado, Simbolo: String): String;
begin
  Result := Estado + '|' + Simbolo;
end;

function ConstruirTabelaTransicoes(const A: TAutomato): TTabela;
var
  Trans: TTransicao;
  Chave: String;
  Destinos: TStringSet;
begin
  Result := TTabela.Create;
  
  for Trans in A.Transicoes do
  begin
    Chave := ChaveTransicao(Trans.Origem, Trans.Simbolo);
    
    if not Result.TryGetValue(Chave, Destinos) then
    begin
      Destinos := TStringSet.Create;
      Result.Add(Chave, Destinos);
    end;
    
    AdicionarAoConjunto(Destinos, Trans.Destino);
  end;
end;

function FechoLambda(Estados: TStringSet; Tabela: TTabela): TStringSet;
var
  Pilha: TStringSet;
  Estado, Proximo: String;
  Chave: String;
  Destinos: TStringSet;
begin
  Result := CopiarConjunto(Estados);
  Pilha := CopiarConjunto(Estados);
  
  try
    while Pilha.Count > 0 do
    begin
      Estado := Pilha[Pilha.Count - 1];
      Pilha.Delete(Pilha.Count - 1);
      
      Chave := ChaveTransicao(Estado, '&');
      
      if Tabela.TryGetValue(Chave, Destinos) then
      begin
        for Proximo in Destinos do
        begin
          if not ConjuntoContem(Result, Proximo) then
          begin
            AdicionarAoConjunto(Result, Proximo);
            Pilha.Add(Proximo);
          end;
        end;
      end;
    end;
  finally
    Pilha.Free;
  end;
end;

function AceitaPalavra(const A: TAutomato; const Palavra: String): Boolean;
var
  Tabela: TTabela;
  EstadosAtuais, ProximosEstados: TStringSet;
  Simbolo: Char;
  Estado, Proximo: String;
  Chave: String;
  Destinos: TStringSet;
  I: Integer;
begin
  Result := False;
  Tabela := ConstruirTabelaTransicoes(A);
  EstadosAtuais := FechoLambda(A.EstadosIniciais, Tabela);
  
  try
    for I := 1 to Length(Palavra) do
    begin
      Simbolo := Palavra[I];
      
      if not ConjuntoContem(A.Alfabeto, Simbolo) then
      begin
        EstadosAtuais.Free;
        LiberarTabela(Tabela);
        Exit(False);
      end;
      
      ProximosEstados := TStringSet.Create;
      
      for Estado in EstadosAtuais do
      begin
        Chave := ChaveTransicao(Estado, Simbolo);
        
        if Tabela.TryGetValue(Chave, Destinos) then
        begin
          for Proximo in Destinos do
            AdicionarAoConjunto(ProximosEstados, Proximo);
        end;
      end;
      
      EstadosAtuais.Free;
      
      if ProximosEstados.Count = 0 then
      begin
        ProximosEstados.Free;
        LiberarTabela(Tabela);
        Exit(False);
      end;
      
      EstadosAtuais := FechoLambda(ProximosEstados, Tabela);
      ProximosEstados.Free;
    end;
    
    for Estado in EstadosAtuais do
      if ConjuntoContem(A.EstadosFinais, Estado) then
      begin
        Result := True;
        Break;
      end;
  finally
    EstadosAtuais.Free;
    LiberarTabela(Tabela);
  end;
end;

procedure LiberarTabela(var T: TTabela);
var
  Conjunto: TStringSet;
begin
  for Conjunto in T.Values do
    Conjunto.Free;
  T.Free;
end;

end.

File: src/automato.py
import json
from typing import List, Set, Dict, Tuple

class Automato:
    
    def __init__(self, alfabeto: List[str], estados: List[str], 
                 estados_iniciais: List[str], estados_finais: List[str],
                 transicoes: List[List[str]]):
        self.alfabeto = set(alfabeto)
        self.estados = set(estados)
        self.estados_iniciais = set(estados_iniciais)
        self.estados_finais = set(estados_finais)
        self.transicoes = transicoes
        
    @classmethod
    def carregar_json(cls, caminho: str):
        with open(caminho, 'r', encoding='utf-8') as f:
            dados = json.load(f)
        return cls(
            dados['alfabeto'],
            dados['estados'],
            dados['estados_iniciais'],
            dados['estados_finais'],
            dados['transicoes']
        )
    
    def salvar_json(self, caminho: str):
        dados = {
            'alfabeto': sorted(list(self.alfabeto)),
            'estados': sorted(list(self.estados)),
            'estados_iniciais': sorted(list(self.estados_iniciais)),
            'estados_finais': sorted(list(self.estados_finais)),
            'transicoes': self.transicoes
        }
        with open(caminho, 'w', encoding='utf-8') as f:
            json.dump(dados, f, indent=4, ensure_ascii=False)     
        
    def construir_tabela_transicoes(self) -> Dict[Tuple[str, str], Set[str]]:
        tabela = {}
        for origem, destino, simbolo in self.transicoes:
            chave = (origem, simbolo)
            if chave not in tabela:
                tabela[chave] = set()
            tabela[chave].add(destino)
        return tabela
    
    def fecho_lambda(self, estados: Set[str], tabela: Dict) -> Set[str]:
        fecho = set(estados)
        pilha = list(estados)
        
        while pilha:
            estado = pilha.pop()
            if (estado, '&') in tabela:
                for proximo in tabela[(estado, '&')]:
                    if proximo not in fecho:
                        fecho.add(proximo)
                        pilha.append(proximo)
        
        return fecho
    
    def aceita_palavra(self, palavra: str, eh_afd: bool = False) -> bool:
        tabela = self.construir_tabela_transicoes()
        estados_atuais = self.fecho_lambda(self.estados_iniciais, tabela)
        
        for simbolo in palavra:
            if simbolo not in self.alfabeto:
                return False
            
            proximos_estados = set()
            
            for estado in estados_atuais:
                if (estado, simbolo) in tabela:
                    proximos_estados.update(tabela[(estado, simbolo)])
            
            if not proximos_estados:
                return False
            
            estados_atuais = self.fecho_lambda(proximos_estados, tabela)
        
        return bool(estados_atuais & self.estados_finais)
